contracts/zero-ex/contracts/src/features/TransformERC20.sol

```diff
@@ -0,0 +1,365 @@
+    function _transformERC20Private(
+        bytes32 callDataHash,
+        address payable taker,
+        IERC20TokenV06 inputToken,
+        IERC20TokenV06 outputToken,
+        uint256 inputTokenAmount,
+        uint256 minOutputTokenAmount,
+        Transformation[] memory transformations
+    )
+        public
+        payable
+        returns (uint256 outputTokenAmount)
+    {
+        // If the input token amount is -1, transform the taker's entire
+        // spendable balance.
+        if (inputTokenAmount == uint256(-1)) {
+            inputTokenAmount = ITokenSpender(address(this))
+                .getSpendableERC20BalanceOf(inputToken, taker);
+        }
+
+        IPuppet puppet = getTransformPuppet();
+
+        // Remember the initial output token balance of the taker.
+        uint256 takerOutputTokenBalanceBefore = minOutputTokenAmount == 0
+            ? 0 : LibERC20Transformer.getTokenBalanceOf(outputToken, taker);
+
+        // Pull input tokens from the taker to the puppet and transfer attached ETH.
+        _transferInputTokensAndAttachedEth(inputToken, taker, address(puppet), inputTokenAmount);
+
+        // Perform transformations.
+        for (uint256 i = 0; i < transformations.length; ++i) {
+            _executeTransformation(puppet, transformations[i], taker, callDataHash);
+        }
+
+        // Ensure enough output token has been sent to the taker.
+        if (minOutputTokenAmount != 0) {
+            // Compute how much output token has been transferred to the taker.
+            uint256 takerOutputTokenBalanceAfter =
+                LibERC20Transformer.getTokenBalanceOf(outputToken, taker);
+            if (takerOutputTokenBalanceAfter > takerOutputTokenBalanceBefore) {
+                outputTokenAmount = takerOutputTokenBalanceAfter.safeSub(
+                    takerOutputTokenBalanceBefore
+                );
+            }
+            if (outputTokenAmount < minOutputTokenAmount) {
+                LibTransformERC20RichErrors.IncompleteERC20TransformError(
+                    address(outputToken),
+                    outputTokenAmount,
+                    minOutputTokenAmount
+                ).rrevert();
+            }
+        }
+
+        // Emit an event.
+        emit TransformedERC20(
+            taker,
+            address(inputToken),
+            address(outputToken),
+            inputTokenAmount,
+            outputTokenAmount
+        );
+    }
+
+    /// @dev Return the current puppet instance that will serve as the execution
+    ///      context for transformations.
+    /// @return puppet The puppet instance.
+    function getTransformPuppet()
+        public
+        override
+        view
+        returns (IPuppet puppet)
+    {
+        return LibTransformERC20Storage.getStorage().puppet;
+    }
+
+    /// @dev Transfer input tokens from the taker and any attached ETH to `to`
+    /// @param inputToken The token to pull from the taker.
+    /// @param taker The taker address.
```

Comment:
It's sort of confusing having a `taker` in this context. Perhaps just `from` and `to`?